tests/client/test_async_client.py

```diff
@@ -166,3 +167,38 @@ async def test_100_continue(server):
 
     assert response.status_code == 200
     assert response.content == data
+
+
+@pytest.mark.usefixtures("async_environment")
+async def test_context_managed_transport():
+    class Transport(httpcore.AsyncHTTPTransport):
+        def __init__(self):
+            self.events = []
+
+        async def aclose(self):
+            self.events.append("transport.aclose")
+
+        async def __aenter__(self):
+            await super().__aenter__()
+            self.events.append("transport.__aenter__")
+
+        async def __aexit__(self):
+            await super().__aexit__()
+            self.events.append("transport.__aexit__")
+
+    # Note that we're including 'proxies' here to *also* run through the
+    # proxy context management, although we can't easily test that at the
+    # moment, since we can't add proxies as transport instances.
+    #
+    # Once we have a more generalised Mount API we'll be able to remove this
+    # in favour of ensuring all mounts are context managed, which will
+    # also neccessarily include proxies.
+    transport = Transport()
+    async with httpx.AsyncClient(transport=transport, proxies="http://www.example.com"):
+        pass
+
+    assert transport.events == [
+        "transport.__aenter__",
+        "transport.aclose",
+        "transport.__aexit__",
+    ]
```

Comment:
I'm curious why `transport.aclose` is included at all here? Do we call `.aclose()` anywhere before `Client.__aexit__()`?

Edit: hmm, probably because the base `.__aexit__()` implementation calls into `.aclose()`?

Reply:
Yes.

The base implementation is "Just call `.close()`/`.aclose()`".

One of the allowable things for a concrete transport to do is "Just implement `close`/`aclose`", which is in fact what our existing implementations do.

That gives us both the simple case, while still allowing *some* implementations to do more complex stuff with `__aenter__`, `__aexit__`.

Reply:
Actually maybe worth a comment there against the test case.